onboarding/mlsys/1-gradients.md

```diff
@@ -0,0 +1,494 @@
+first `(x, ytrue)` pair of your dataset.
+
+Here are some test cases:
+
+```python
+dataset = [(1,1)]
+mfinal = train_loop(m=0, lr=0.001, dataset=dataset, iterations=1)
+assert mfinal == 0.002, mfinal
+
+dataset = [(1,1)]
+mfinal = train_loop(m=0, lr=0.001, dataset=dataset, iterations=100)
+assert round(mfinal, 5) == 0.18143, mfinal
+
+
+dataset = [(1,1),(2,2),(3,3),(4,4)]
+mfinal = train_loop(m=0, lr=0.001, dataset=dataset, iterations=4)
+assert round(mfinal, 5) == 0.05891, mfinal
+
+
+dataset = [(1,1),(2,2),(3,3),(4,4)]
+mfinal = train_loop(m=0, lr=0.001, dataset=dataset, iterations=100)
+assert round(mfinal, 5) == 0.78086, mfinal
+```
+
+### Q7
+
+What happens if you set the learning rate way too low?
+
+```python
+dataset = [(1,2),(2,6),(3,7),(4,4)]
+mfinal = train_loop(m=0, lr=0.000000001, dataset=dataset, iterations=4)
+print(mfinal)
+```
+
+Explain in words what is happening.
+
+### Q8
+
+What happens if you set the learning rate way too high?
+
+```python
+dataset = [(1,1),(2,2),(3,3),(4,4)]
+mfinal = train_loop(m=0, lr=10, dataset=dataset, iterations=4)
+print(mfinal)
+```
+
+Explain in words what is happening.
+
+## Noise in Training
+
+You should have a pretty good grasp of how the gradient "steers" the training
+of the model towards better predictions over time.  We've been using an
+artificially easy dataset to make it obvious what the ideal value for `m` would
+be, but let's change up the dataset to make it more noisy.  The ideal value of
+`m` will still be 1, the loss just won't ever go to zero anymore.
+
+Our new dataset will be:
+
+- `(x=1, ytrue=0)`
+- `(x=2, ytrue=1)`
+- `(x=3, ytrue=1)`
+- `(x=4, ytrue=6.25)`
+
+You can check that the model still should converge to `1` either by solving the
+loss function algebraically or running your training loop:
+
+```python
+dataset = [(1,0),(2,1),(3,1),(4,6.25)]
+print(train_loop(m=0, lr=0.0001, dataset=dataset, iterations=10000))
+```
+
+Now that we have noise in our dataset, we find that some values are "noisier"
+than others.  Notice that the at `m=1` our model is off by `1`, but at `x=4`
+our model is off by `1.5`.  These noisier points can cause training to become
+unstable at higher learning rates.
+
+### Q9
+
+Test your `train_loop` function against different learning rates:
+
+```python
+dataset = [(1,0),(2,1),(3,1),(4,6.25)]
+for lr in [.001, .002, .004, .008, .016, .032, .064, .128, .256]:
+    print("lr", lr, train_loop(m=0, lr=lr, dataset=dataset, iterations=1000))
+```
+
+Which learning rates cause training to diverge?  Which learning rates cause the
+model to converge to 1?  Are there values which are somewhere in between?
+
+## Batch size in Training
+
+In real life, all datasets have noise.  If you calculate the gradient for every
+record and step the model on each data point's gradient, that noise can cause
+you to take a lot of wrong steps, causing training to either take longer to
+converge or not to converge at all.
+
```

Comment:
....to either take longer to converge or to not converge at all.

Reply:
... ~to~ not converge ...

The "to" is already present before the "either", so it should not be repeated within the list.